### ISSUE TITLE
Container garbage collection should define MinAge as age since the container termination

### ISSUE THREAD

**Issue by yujuhong** [kubernetes/kubernetes#13562]
The definition of `MinAge` in Container GC:

```
// Minimum age at which a container can be garbage collected, zero for no limit.
MinAge time.Duration
```

It determines a container's age by looking at its "Created" time. Even if a container has just died, as long as the container was created `MinAge` before, it would be eligible to be GC'd.
This seems arbitrary to me. What we should be looking at is "how long since the container died", so that other components in kubelet has a chance to look at the dead container before it's GC'd.

/cc @vishh @pwittrock

**Comment by vishh** [comment]
What are the use cases for keeping old containers? AFAIK, we use old
containers for tracking restart counts and providing access to previous
logs. If we can avoid requiring old containers, we can greatly improve disk
utilization.
A container that writes a lot of data to its rootfs causes lots of disk
usage issues. We should try to GC those containers as quickly as possible.

**Comment by yujuhong** [comment]
> What are the use cases for keeping old containers? AFAIK, we use old
> containers for tracking restart counts and providing access to previous
> logs. If we can avoid requiring old containers, we can greatly improve disk
> utilization.

Even if we write restart counts as labels to the containers, we'd still need to keep a most recently terminated container around before we create a new one (so that we can read the label back).
Reading log is the same situation. The current GC parameters does not guarantee that at all.

> A container that writes a lot of data to its rootfs causes lots of disk
> usage issues. We should try to GC those containers as quickly as possible.

Agreed. I'd imagine that the recent per-container backoff limits the rate of generating dead containers, so this would relieve the problem a bit. To actually solve this, we need to collect the data from the dead container and mark it eligible for GC right afterwards.

**Comment by vishh** [comment]
On Thu, Sep 3, 2015 at 10:42 AM, Yu-Ju Hong <notifications@github.com>
wrote:

> What are the use cases for keeping old containers? AFAIK, we use old
> containers for tracking restart counts and providing access to previous
> logs. If we can avoid requiring old containers, we can greatly improve disk
> utilization.
>
> Even if we write restart counts as labels to the containers, we'd still
> need to keep a most recently terminated container around before we create a
> new one (so that we can read the label back).
> Reading log is the same situation. The current GC parameters does not
> guarantee that at all.
>
Good point.

> A container that writes a lot of data to its rootfs causes lots of disk
> usage issues. We should try to GC those containers as quickly as possible.
>
> Agreed. I'd imagine that the recent per-container backoff limits the rate
> of generating dead containers, so this would relieve the problem a bit. To
> actually solve this, we need to collect the data from the dead container
> and mark it eligible for GC right afterwards.
>
+1 for marking dead containers for GC.

**Comment by yujuhong** [comment]
> +1 for marking dead containers for GC.

Ideally, the pod lifecycle event generator would detect the container change, trigger a cache update, and then mark the container GC-eligible.
However, to survive kubelet restarts, we'd still need to keep at least *one* container per container name in the pod

**Comment by samsabed** [comment]
@yujuhong Is this a wish list or something you are working on ? I can look into it.
> However, to survive kubelet restarts, we'd still need to keep at least one container per container name in the pod
Is that because of how GetPodStatus works in docker ?

**Comment by yujuhong** [comment]
> @yujuhong Is this a wish list or something you are working on ? I can look into it.

I am not working on this. It'd be great if you want to take a look :) @vishh is working on disk management, and he may have more plans on GC.

> Is that because of how GetPodStatus works in docker ?

This is because kubelet doesn't persist any state, so we rely on dead containers on state. We'd probably need to keep one dead container so that we can know the restart count.

**Comment by samsabed** [comment]
@yujuhong @vishh marking the containers can be done either as a label (docker) which would make the restart count as labels similar though would leave rkt unimplemented till meta-service lands. or can be in memory  (statusManage ?) which would make it work for rkt but diverge for restart.
WDYT

**Comment by yujuhong** [comment]
> @yujuhong @vishh marking the containers can be done either as a label (docker) which would make the restart count as labels similar though would leave rkt unimplemented till meta-service lands. or can be in memory (statusManage ?) which would make it work for rkt but diverge for restart.
> WDYT

@samabed, rkt already implemented this by writing the restart count to the service file. Using docker label would be similar to that.

We do use statusManager to keep track of the restart count for docker. The downside is that, it still needs to find out "how many containers have died since the last check", and increment the restart count accordingly. Writing restart count as a label eliminate this need, as kubelet would only need to look at the most recently died container, increment the count, and write it to the new container that's being created.

**Comment by Random-Liu** [comment]
I just put restart count into docker label, although the commit is still blocked by shippable test...

For this problem, I have several questions:
  1. How could garbage collector get the "GC-eligible mark"? The garbage collector directly get container information from docker client now. It seems that now we can't make some mark on a started docker container, right?

  2. Now we retain restart count in docker label, so after a dead container is restarted, we'll never need it any more for restart count. However, it is said that dead container is also kept for previous log. If so, when can I mark a container "GC-eligible"?

If these two questions solved, I think I can try to add the "GC-eligible mark" now. :)

**Comment by Random-Liu** [comment]
Some humble thoughts for the first question:
    * Let garbage collector have a reference of status manager?
    * Save some shared data in garbage collector?
    * Create a channel, and send updates to garbage collector?

**Comment by yujuhong** [comment]
Good points.

The discussions in this thread are still at the early stage. We don't have a good way to mark a container GC-eligible now and some investigation is needed.
 - In the long run, we are moving to the client-server architecture for container runtimes (#13768), so it's good to keep in mind about compatibility. I'd avoid sharing data directly with the garbage collector in the container runtime.
 - @dchen1107 wants a disk manager to handle all disk space issues -- including containers and images. There is no detailed proposal yet.

As for what containers are eligible for GC:
 1. containers in a deleted pods are eligible.
 2. Except for the most recent two dead containers, any container (with the same container name) in a pod can be GC'd.

In the meantime, I think `MinAge` should still be fixed to improve the reliability of generating pod status.

**Comment by yujuhong** [comment]
> Some humble thoughts for the first question:
> * Let garbage collector have a reference of status manager?
> * Save some shared data in garbage collector?
> * Create a channel, and send updates to garbage collector? (Consistent with the future event-driven model)

We can have a high-level, runtime-agnostic garbage collector that receives the updates from the channel, and calls the runtime-provided garbage collection function to either 1) perform generic GC, or 2) delete specific pods. 
This would allow GC to be more aggressive, preventing too many dead containers from downgrading the docker performance (as suggested by your benchmarking results).

**Comment by vishh** [comment]
GC is closely related to Disk accounting. I'd prefer typing them together.
The GC logic should derive the state of the system from a central location
though.

**Comment by Random-Liu** [comment]
1) Now garbage collector also calls List and Inspect to get the status of containers, which will have already been cached in runtime pod cache (we are going to implement it). It seems that garbage collector doesn't need to access the runtime again.
2) If we want to mark something like GC-eligible, we have to do that in pod management components such as PLEG.
So at least, garbage collector should receive some information from PLEG, right?

**Comment by yujuhong** [comment]
> GC is closely related to Disk accounting. I'd prefer typing them together.
> The GC logic should derive the state of the system from a central location
> though.

@vishh, that was what I thought initially, but now I think container GC may be different from the rest (e.g., image). 
 - There is no value keeping dead containers of deleted and/or terminated pods. This is different from, say, images, which you may be able to reuse later.
 - For pods that are still alive, we know exactly how many dead containers is required to generate pod statuses. Beyond that, the containers serve no value.

To sum up, we know what containers can be removed exactly, and keeping them around does us no good, other than degrading docker's performance. On top of that, when there is disk space pressure, you can't really remove *more* dead containers and risk corrupting the pod status.

**Comment by vishh** [comment]
> There is no value keeping dead containers of deleted and/or terminated
pods. This is different from, say, images, which you may be able to reuse
later.

This is not entirely true. We provide access to previous instance logs. I'm
trying to find ways where logs are independent of container lifetimes
though.

> To sum up, we know what containers can be removed exactly, and keeping
them around does us no good, other than degrading docker's performance. On
top of that, when there is disk space pressure, you can't really remove
*more* dead containers and risk corrupting the pod status.

With the labels change, we will stop relying on tombstones right or do we
still have use cases for tombstones?

**Comment by Random-Liu** [comment]
@vishh, it seems that the latest dead container is still useful, before the new container is started.

**Comment by vishh** [comment]
Yes. On a best-effort basis ideally.

**Comment by yujuhong** [comment]
> This is not entirely true. We provide access to previous instance logs. I'm
> trying to find ways where logs are independent of container lifetimes
> though.

Yes, let me rephrase that. For *deleted* pods, we no longer have to keep the dead containers.

> With the labels change, we will stop relying on tombstones right or do we
> still have use cases for tombstones?

As @Random-Liu pointed out, you still need latest dead container. Moreover, we still support `kubectl logs -p`, so one previously dead container is required until we can resolve all the issues.

*If* we can stop really on tombstones completely, there is even more reason to GC all the dead containers ASAP, regardless of the current disk usage :-)

**Comment by vishh** [comment]
Previous logs are best-effort. So if the node only has a 10GB disk, we
should go ahead and delete previous container instances.

**Comment by vishh** [comment]
In general, I agree with the overall idea of being aggressive about garbage
collecting dead containers.

**Comment by yujuhong** [comment]
The issue is no longer relevant now that the GC has changed significantly.
